src/query/executor/request.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package executor
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+	"github.com/m3db/m3/src/query/plan"
+	"github.com/m3db/m3/src/query/util/logging"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+// State is the request state.
+type State int
+
+const (
+	compiling State = iota
+	planning
+	executing
+	all
+)
+
+func (s State) String() string {
+	switch s {
+	case compiling:
+		return "compiling"
+	case planning:
+		return "planning"
+	case executing:
+		return "executing"
+	case all:
+		return "all"
+	default:
+		return "unknown"
+	}
+}
+
+func (s State) durationString() string {
+	return fmt.Sprint("%s_duration_seconds", s)
+}
+
+// Request represents a single request.
+type Request struct {
+	engine     *Engine
+	params     models.RequestParams
+	parentSpan *span
+}
+
+func newRequest(ctx context.Context, engine *Engine, params models.RequestParams) (*Request, context.Context) {
+	parentSpan, parentCtx := startSpan(ctx, all.String(), engine.metrics.activeHist, engine.metrics.all)
+	r := &Request{engine: engine, params: params, parentSpan: parentSpan}
+	return r, parentCtx
+
+}
+
+func (r *Request) compile(ctx context.Context, parser parser.Parser) (parser.Nodes, parser.Edges, error) {
+	sp, _ := startSpan(ctx, compiling.String(), r.engine.metrics.compilingHist, r.engine.metrics.compiling)
+	// TODO: Change DAG interface to take in a context
+	nodes, edges, err := parser.DAG()
+	if err != nil {
```

Comment:
if this err occurs, the span won't be finished. Might be better to defer `sp.finish()`. Ditto in the other 2 functions below.

Reply:
also, you could do something like:

```
import (
"github.com/opentracing/opentracing-go/ext"
"github.com/opentracing/opentracing-go/log"
)

defer func() {
   ext.Error.Set(sp.s, true)
   sp.s.LogFields(log.Error(err))
   sp.finish()
}
```

There are also database specific opentracing tags: https://github.com/opentracing/opentracing-go/blob/master/ext/tags.go#L105